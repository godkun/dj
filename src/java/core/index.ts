/**
 * 通过构造Java Style code的DSL来自动生成 ts，java， go的dubbo框架的调用代码
 */

import Clazz from './clazz'
import Service from './service'
import {
  Integer,
  String,
  Boolean,
  List,
  Set,
  Collection,
  Map,
  boolean,
  int,
  short,
  Short,
  byte,
  Byte,
  long,
  Long,
  double,
  Double,
  float,
  Float,
  char,
  Char,
  chars,
  Character,
  Enumeration,
  Iterator,
  Currency,
  Dictionary,
  HashMap,
} from './java-type'

import Enum from './enum'

export default {
  Enum,
  Clazz,
  Service,

  Boolean,
  boolean,

  Integer,
  int,

  short,
  Short,

  byte,
  Byte,

  long,
  Long,

  double,
  Double,

  float,
  Float,

  String,
  char,
  Char,
  chars,
  Character,

  List,
  Set,
  Collection,
  Iterator,
  Enumeration,
  Map,
  HashMap,
  Dictionary,
  Currency,
}
