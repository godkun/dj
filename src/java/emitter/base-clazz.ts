import path from 'path'

import debug from 'debug'
import Deps from '../core/deps'
import fsx from 'fs-extra'
import fmt from './fmt'

const log = debug(`dubbo:dj:emitter/BaseClazzï¼šsay ~`)
const relWhiteList = ['js-to-java', '@dubbo/dj-suger']

const fileType = {
  ts: 'ts',
  go: 'go',
  java: 'java',
}

export default abstract class BaseClazz {
  curFilePath: string

  constructor(
    public fullClsName: string,
    lang: 'ts' | 'go' | 'java',
    baseDir: string
  ) {
    this.curFilePath = `${baseDir}/${lang}/${this.cls2Path(fullClsName)}.${
      fileType[lang]
    }`
  }

  imports(deps: Deps) {
    const imports = []
    for (let { fullClassName, importName } of deps.imports) {
      if (
        relWhiteList.includes(fullClassName) ||
        fullClassName.startsWith('.')
      ) {
        imports.push(`import ${importName} from '${fullClassName}'`)
      } else {
        let rel = this.relPath(this.fullClsName, fullClassName)
        if (!rel.startsWith('./') && !rel.startsWith('../')) {
          rel = './' + rel
        }
        imports.push(`import ${importName} from '${rel}'`)
      }
    }
    return imports
  }

  get mott() {
    return `
      /**
       * auto generated by dubbo dj
       * ~~~ ðŸ’— mechine coding ðŸ’— ~~~
       */
    `
  }

  cls2Path(cls: string) {
    return cls.replace(/\./g, '/')
  }

  relPath(from: string, to: string) {
    const fromArr = from.split('.')
    fromArr.pop()
    // è®¡ç®—ç›¸å¯¹è·¯å¾„
    const relPath = path.relative(fromArr.join('/'), this.cls2Path(to))
    log(`${from} => relative => ${to} =>`, relPath)
    return relPath
  }

  // write to => ./dubbo/ts/**/*.ts
  async writeCode(): Promise<void> {
    try {
      await fsx.ensureFile(this.curFilePath)
      await fsx.writeFile(this.curFilePath, this.pretteyCode)
      return log(
        `create ${this.fullClsName} => ${this.curFilePath} was successfully`
      )
    } catch (err) {
      log(`generate => '${this.fullClsName}' was failed, err:` + err)
      console.error(err)
    }
  }

  get pretteyCode(): string {
    return fmt(this.code)
  }

  abstract get code(): string
}
