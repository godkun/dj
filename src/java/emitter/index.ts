import Clazz from './clazz'
import Enum from './enum'
import Service from './service'
import AbstractService from './service-abstract'

export { Enum, Clazz, Service, AbstractService }
