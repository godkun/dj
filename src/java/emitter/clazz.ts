import JavaBase from './base-clazz'
import { JavaClazz } from '../core/clazz'
import debug from 'debug'
import { clazzDot } from './dot/ts'

const log = debug(`dubbo:dj:emitter/clazzï¼šsay ~`)

export default class Clazz extends JavaBase {
  constructor(
    private clazz: JavaClazz,
    lang: 'ts' | 'go' | 'java',
    private baseDir: string = './dubbo'
  ) {
    super(clazz.fullClsName, lang, baseDir)
  }

  get code(): string {
    return clazzDot({
      mott: this.mott,
      imports: this.imports(this.clazz.deps).join(';'),
      cls: this.clazz.fullClsName,
      infName: this.clazz.infName,
      fields: this.clazz.fields,
      clsName: this.clazz.clsName,
    })
  }
}
