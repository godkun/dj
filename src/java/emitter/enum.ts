import { JavaEnum } from '../core/enum'
import JavaBase from './base-clazz'
import { enumDot } from './dot/ts'

export default class Enum extends JavaBase {
  constructor(
    public javaEnum: JavaEnum,
    lang: 'ts' | 'go' | 'java',
    baseDir: string = './dubbo'
  ) {
    super(javaEnum.fullClsName, lang, baseDir)
  }

  get code() {
    return enumDot({
      mott: this.mott,
      name: this.javaEnum.clsName,
      fields: this.javaEnum.fields,
    })
  }
}
