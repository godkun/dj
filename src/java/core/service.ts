import { combine } from 'js-to-java'
import { IType, TServiceMethod } from '../types'
import BaseClazz from './base-clazz'
import { JavaClazz } from './clazz'

export class JavaService extends BaseClazz {
  _group: string = ''
  _version: string = '1.0.0'
  _methods: TServiceMethod = {}
  private _curMethodName: string = ''

  constructor(fullClsName: string) {
    super(fullClsName)
  }

  group(group: string) {
    this._group = group
    return this
  }

  version(version: string) {
    this._version = version
    return this
  }

  method(name: string) {
    // set current method
    this._curMethodName = name

    // init method
    this._methods[name] = {
      args: [],
      ret: null,
    } as any

    return this
  }

  arg(name: string, type: JavaClazz | (() => IType)) {
    const method = this._methods[this._curMethodName]

    if (type instanceof JavaClazz) {
      const renamed = this.deps.add(type.fullClsName, type.clsName)
      method.args.push({
        name,
        type: renamed,
      })
    } else {
      // FIXEDME 泛型的处理
      method.args.push({
        name,
        type: type().tsType,
      })
    }

    return this
  }

  ret(type: JavaClazz | (() => IType)) {
    // get current method meta data
    const method = this._methods[this._curMethodName]

    if (type instanceof JavaClazz) {
      const renameClsName = this.deps.add(type.fullClsName, type.clsName)
      method.ret = renameClsName
    } else {
      // FXIEDME 反省处理
      method.ret = type().tsType
    }

    return this
  }
}

export default function Service(fullClsName: string) {
  const service = new JavaService(fullClsName)
  return {
    group(group: string) {
      service.group(group)
      return this
    },

    version(version: string) {
      service.version(version)
      return this
    },

    method(name: string) {
      service.method(name)
      return this
    },

    arg(name: string, type: JavaClazz | (() => IType)) {
      service.arg(name, type)
      return this
    },

    ret(type: JavaClazz | (() => IType)) {
      service.ret(type)
      return this
    },
    ok() {
      return service
    },
  }
}
