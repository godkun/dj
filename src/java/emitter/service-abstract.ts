import JavaBase from './base-clazz'
import { JavaService } from '../core/service'
import { abstractServiceDot } from './dot/ts'

export default class AbstractService extends JavaBase {
  constructor(
    public service: JavaService,
    lang: 'ts' | 'go' | 'java',
    baseDir: string = './dubbo'
  ) {
    super(service.fullClsName, lang, baseDir)
    this.curFilePath = this.curFilePath.replace(
      service.clsName,
      `base/${service.clsName}`
    )
    this.fullClsName = this.fullClsName.replace(
      service.clsName,
      `base.${service.clsName}`
    )
  }

  get code() {
    const methods = []
    const args = []

    for (let [name, meta] of Object.entries(this.service._methods)) {
      for (let arg of meta.args) {
        args.push(`${arg.name}: ${arg.type}`)
      }
      methods.push(
        `abstract ${name}(${args.join()})${meta.ret ? ':' + meta.ret : 'void'};`
      )
    }

    return abstractServiceDot({
      mott: this.mott,
      imports: this.imports(this.service.deps).join(';'),
      service: this.service.fullClsName,
      serviceName: this.service.clsName,
      group: this.service._group,
      version: this.service._version,
      methods: methods,
    })
  }
}
