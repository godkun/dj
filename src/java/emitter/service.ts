import fsx, { exists } from 'fs-extra'
import JavaBase from './base-clazz'
import { JavaService } from '../core/service'
import { serviceDot } from './dot/ts'
import debug from 'debug'

const log = debug(`dubbo:dj:service:say ~`)

export default class Service extends JavaBase {
  constructor(
    public service: JavaService,
    lang: 'ts' | 'go' | 'java',
    baseDir: string = './dubbo'
  ) {
    super(service.fullClsName, lang, baseDir)
  }

  async writeCode(): Promise<void> {
    // 如果存在就不覆盖
    const exists = await fsx.pathExists(this.curFilePath)
    if (!exists) {
      super.writeCode()
    } else {
      log(`${this.curFilePath} was existed, skip it ~~~~~~`)
    }
  }

  get code() {
    const methods = []
    const args = []

    for (let [name, meta] of Object.entries(this.service._methods)) {
      for (let arg of meta.args) {
        args.push(`${arg.name}: ${arg.type}`)
      }
      methods.push(`
        ${name}(${args.join()})${meta.ret ? ':' + meta.ret : 'void'} {
          throw new Error('Method not implemented.')
        }
      `)
    }

    return serviceDot({
      mott: this.mott,
      imports: this.imports(this.service.deps).join(';'),
      service: this.service.fullClsName,
      serviceName: this.service.clsName,
      group: this.service._group,
      version: this.service._version,
      methods: methods,
    })
  }
}
