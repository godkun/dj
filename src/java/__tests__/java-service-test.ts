import fmt from '../emitter/fmt'
import java from '../core'
import { Service } from '../emitter'
import { userService } from '../../cmd/dsl/service'

const user = java
  .Clazz('org.apache.dubbo.domain.User')
  .field('id', java.Integer)
  .ok()

const UserService = java
  .Service('org.apache.dubbo.service.UserService')
  .group('A')
  .version('1.0.0')
  .method('sayHello')
  .arg('isExit', java.Boolean)
  .arg('user', user)
  .ret(java.Integer)
  .method('sayHi')
  .arg('i', java.Integer)
  .ret(java.String)
  .ok()

it('test java service', () => {
  expect(JSON.stringify(UserService, null, 2)).toMatchSnapshot()
  expect(new Service(userService, 'ts').pretteyCode).toMatchSnapshot()
})
