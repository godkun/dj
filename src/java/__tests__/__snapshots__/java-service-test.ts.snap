// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test java service 1`] = `
"{
  \\"fullClsName\\": \\"org.apache.dubbo.service.UserService\\",
  \\"clsName\\": \\"UserService\\",
  \\"infName\\": \\"IUserService\\",
  \\"deps\\": {
    \\"map\\": {}
  },
  \\"_group\\": \\"A\\",
  \\"_version\\": \\"1.0.0\\",
  \\"_methods\\": {
    \\"sayHello\\": {
      \\"args\\": [
        {
          \\"name\\": \\"isExit\\",
          \\"type\\": \\"boolean\\"
        },
        {
          \\"name\\": \\"user\\",
          \\"type\\": \\"User\\"
        }
      ],
      \\"ret\\": \\"number\\"
    },
    \\"sayHi\\": {
      \\"args\\": [
        {
          \\"name\\": \\"i\\",
          \\"type\\": \\"number\\"
        }
      ],
      \\"ret\\": \\"string\\"
    }
  },
  \\"_curMethodName\\": \\"sayHi\\"
}"
`;

exports[`test java service 2`] = `
"/**
 * auto generated by dubbo dj
 * ~~~ ðŸ’— mechine coding ðŸ’— ~~~
 */

import User from '../domain/User'
import UserService from './base/UserService'

export default class UserServiceImpl extends UserService {
  dubboInterface = 'org.apache.dubbo.service.UserService'
  group = ''
  version = '1.0.0'

  sayHello(user: User): User {
    throw new Error('Method not implemented.')
  }

  sayWorld(user: User, name: string): string {
    throw new Error('Method not implemented.')
  }
}
"
`;
